dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/RateLimiter.java

```diff
@@ -0,0 +1,224 @@
+    private final long rateLimit;
+    private final Duration rateLimitDuration;
+    private final Duration tokenAddingRate;
+    private final Bucket bucket;
+
+    public RateLimiter() {
+        this(Constants.DEFAULT_RATE_LIMIT, Constants.DEFAULT_RATE_LIMIT_DURATION,
+                Constants.DEFAULT_TOKEN_ADDING_RATE);
+    }
+
+    /**
+     * constructor to initialize RateLimiter with required arguments.
+     * If time meter is specified, the system default will be chosen.
+     *
+     * @param rLimit          maximum count of PRs to be sent per rLimitDuration
+     * @param rLimitDuration  Rate Limit duration
+     * @param tokAddingRate   Rate at which tokens are added in the bucket
+     * @param customTimeMeter Clock to be used for bucketing the tokens.
+     *                        Defaults to TimeMeter.SYSTEM_MILLISECONDS
+     * @param <T>             Implementing class for customTimeMeter must be a class
+     *                        extending TimeMeter
+     * @see TimeMeter
+     * @see TimeMeter#SYSTEM_MILLISECONDS
+     */
+    public <T extends TimeMeter> RateLimiter(long rLimit, Duration rLimitDuration,
+                                             Duration tokAddingRate, T customTimeMeter) {
+        rateLimit = rLimit;
+        rateLimitDuration = rLimitDuration;
+        tokenAddingRate = tokAddingRate;
+        customTimeMeter = customTimeMeter != null ? customTimeMeter : (T) TimeMeter.SYSTEM_MILLISECONDS;
+        // refill the bucket at the end of every 'rateLimitDuration' with 'rateLimit' tokens,
+        // not exceeding the max capacity
+        Refill refill = Refill.intervally(rateLimit, rateLimitDuration);
+        // initially bucket will have no. of tokens equal to its max capacity i.e
+        // the value of 'rateLimit'
+        Bandwidth limit = Bandwidth.classic(rateLimit, refill);
+        this.bucket = Bucket.builder()
+                .addLimit(limit)
+                // this is internal limit to avoid spikes and distribute the load uniformly over
+                // DEFAULT_RATE_LIMIT_DURATION
+                // one token added per DEFAULT_TOKEN_ADDING_RATE
+                .addLimit(Bandwidth.classic(1, Refill.intervally(1, tokenAddingRate)))
+                .withCustomTimePrecision(customTimeMeter)
+                .build();
+    }
+
+    public RateLimiter(long rateLimit, Duration rateLimitDuration,
+                       Duration tokenAddingRate) {
+        this(rateLimit, rateLimitDuration, tokenAddingRate, null);
+
+    }
+
+    /**
+     * This method will create and return an object of type RateLimiter based on the
+     * variable set in the Namespace.
+     *
+     * @param ns Namespace {@see net.sourceforge.argparse4j.inf.Namespace}
+     * @return RateLimiter object if required variable is set in Namespace
+     *          null otherwise.
+     * @see net.sourceforge.argparse4j.inf.Namespace Namespace
+     */
+    public RateLimiter getRateLimiter(Namespace ns) {
```

Comment:
This can be static correct?  Looks like a factory method.  Can we return an Optional<RateLimiter> instead of returning null?